File: backend/src/Features/Auth/Infra/Repositories/RegisterUserRepository.php
<?php


namespace Rodri\VotingApp\Features\Auth\Infra\Repositories;


use Exception;
use Rodri\VotingApp\Features\Auth\Domain\Entities\User;
use Rodri\VotingApp\Features\Auth\Domain\Repositories\IRegisterUserRepository;
use Rodri\VotingApp\Features\Auth\Domain\ValueObjects\Email;
use Rodri\VotingApp\Features\Auth\Infra\DataLayer\IRegisterUserDataLayer;
use Rodri\VotingApp\Features\Auth\Infra\DataTransferObjects\UserDTO;
use Rodri\VotingApp\Features\Auth\Infra\Exceptions\RegisterUserDataLayerException;

/**
 * Repository - RegisterUserRepository
 * @package Rodri\VotingApp\Features\Auth\Infra\Repositories
 * @author Rodrigo Andrade
 */
class RegisterUserRepository implements IRegisterUserRepository
{

    /**
     * RegisterUserRepository constructor.
     * @param IRegisterUserDataLayer $dataLayer
     */
    public function __construct(
        private IRegisterUserDataLayer $dataLayer
    )
    {
    }

    /**
     * @param User $user
     * @throws RegisterUserDataLayerException
     */
    public function invoke(User $user): void
    {
        try {
            $this->dataLayer->invoke(UserDTO::factoryUserDTO($user));
        } catch (Exception $e) {
            throw new RegisterUserDataLayerException('Impossible register a user');
        }
    }

    /**
     * @param Email $email
     * @return bool
     * @throws RegisterUserDataLayerException
     */
    public function hasEmailAlready(Email $email): bool
    {
        try {
            return $this->dataLayer->hasEmailAlready($email);
        } catch (Exception $e) {
            throw new RegisterUserDataLayerException('Impossible check if email already exist.');
        }
    }
}

File: app/lib/Features/authentication/infra/repositories/register_user_repository.dart
import 'package:app/Features/authentication/domain/entities/user.dart';
import 'package:app/Features/authentication/infra/models/user_mode.dart';
import 'package:app/Features/authentication/infra/datalayers/abstract_register_user_data_layer.dart';
import 'package:app/Features/authentication/infra/exceptions/register_user_repository_exception.dart';
import 'package:app/Features/authentication/domain/repositories/abstract_register_user_repository.dart';

/// Implementação - AbstractRegisterUserRepository
////
/// author: Rodrigo Andrade
class RegisterUserRepository implements AbstractRegisterUserRepository {
  AbstractRegisterUserDataLayer _dataLayer;

  RegisterUserRepository(this._dataLayer);

  @override
  Future<void> call(User? user) async {
    try {
      _dataLayer(UserModel.by(user));
    } on Exception catch (error) {
      throw RegisterUserRepositoryException(error.toString());
    }
  }
}
